### ISSUE TITLE
Proposal: Expose Applied Filters

### ISSUE THREAD

**Issue by MarkHerhold** [crossfilter/crossfilter#44]
I'm running into an issue where I'd like to be able to determine what filters crossfilter is applying to my data. I don't want to keep track of the filters myself outside of crossfilter because that is an issue with regards to 'separation of concerns' and other basic software development fundamentals and a great opportunity for bugs.

[Related StackOverflow question](http://stackoverflow.com/questions/28354172/how-do-i-get-the-current-filter-for-a-crossfilter-dimension) on the same topic.
#### `Dimension` -> `currentFilter()`

`dimension.currentFilters()` - returns an array of filters applied to the current dimension

``` js
var payments = crossfilter([...]);
var paymentsByProductID = payments.dimension(function(d) { return d.productIDs; }, true);
paymentsByProductID.filter("004");
paymentsByProductID.currentFilter(); // -> "004"
```
#### `crossfilter` -> `currentFilters()`

`crossfilterInstance.currentFilters()` - returns all filters for all dimensions on the crossfilter instance

``` js
var payments = crossfilter([...]);
var paymentsByProductID = payments.dimension(function(d) { return d.productIDs; }, true);
paymentsByProductID.filter("004");
var paymentsByQuantity = payments.dimension(function(d) { return d.quantity; }, true);
paymentsByQuantity.filter(2);

// get the filters on the crossfilter instance
payments.currentFilters();
```

Returns

``` js
[{
    dimension: paymentsByProductID, // the payments dimension instance
    filter: "004"
}, {
    dimension: paymentsByQuantity, // the quantity dimension instance
    filter: 2
}]
```
#### `crossfilter` / `Dimension` -> `on('filter', callbackFn)`

Accepts a callback function that is called when the filter changes on the crossfilter instance or on the dimension. This would be useful to see when something changes what is filtered (e.g. useful for dashboards).

``` js
var payments = crossfilter([...]);
payments.on('filter', function(changed) {
    // where `changed` is the same as the result of payments.currentFilters()
    //  changed -> [{
    //      dimension: paymentsByProductID, // the payments dimension instance
    //      filter: "004"
    //  }]
});
var paymentsByProductID = payments.dimension(function(d) { return d.productIDs; }, true);
paymentsByProductID.filter("004");

// the above callback function would be triggered
```

I'd like to hear others' opinions on this. I don't know how hard this would be to implement, but as an API consumer, this is what I would like to see. Thanks! :beers:

**Comment by gordonwoodhull** [comment]
I agree it's a good idea, and one of our goals in this fork is greater transparency.

A couple nits (please edit your proposal if you agree):
 - there is only one filter per dimension, so that one should be singular `currentFilter`
 - I think you mean that `crossfilterInstance.currentFilters` should return all filters for all dimensions

The more serious objection is that this won't work for [filter functions](https://github.com/crossfilter/crossfilter/wiki/API-Reference#dimension_filterFunction), or it will work but it will return something that is a black box. (All you can do with function objects is call them and turn them into text, and interpreting the text is error-prone.) This might not be a serious issue, though: it just means that the API won't be useful for people who use that functionality.

I'm not fond of event subscriptions on setters, but I can live with that. I'd favor a more idiomatic `.on('filterAdded', ...)` if we do that. I guess my objection is that it seems like a slippery slope to needing subscriptions on every single parameter everywhere.

**Comment by MarkHerhold** [comment]
@gordonwoodhull I agree with your bullet points. I'll update my proposal shortly.

Regarding filter functions, I don't see why we can't just return a reference to the filter function itself. This way, users can do whatever they want with the function including use it for other crossfilter filter calls, turn it into text, mutate it (not actually recommended), etc etc. Basically I want to leave the decision on what to do with it up to the user. If the filter was on a number, we'll return the number. Likewise for strings and functions. I don't think it should make a difference.

Regarding the `on()` instead of `subscribe()`: My only thought is that `.on()` normally indicates that an event has been fired (e.g. click DOM event). `.subscribe()` doesn't necessarily indicate an "event" to me and would fit better with the node ecosystem in my opinion. I don't have a strong opinion on what we name it though and am fine with either.

On the same `subscribe()/on()` note, I think we should structure the strings like so:

- `.on('filter')` - triggered for any filter change
- `.on('filter:add')` - triggered for a filter being added
- `.on('filter:remove')` - triggered for a filter being removed

Alternatively, we can wrap this information in an event object and pass it as the first argument to the callback, keeping the API simple.

**Comment by gordonwoodhull** [comment]
I don't see why they are not events just because these are not DOM elements. Is `subscribe` a Node convention? 

I haven't seen the `:` syntax for events, either, so it's probably my unfamiliarity. I guess most of my experience in Javascript has been around d3.

**Comment by esjewett** [comment]
I agree this is basically a good idea. Shouldn't be hard to implement either.

Regarding the subscribe callback, please take a look at the currently undocumented (paging @tannerlinsley ;-) ) crossfilter.onChange, which will notify of filter updates. If you want to add functionality to that regarding filters, I think that would be fine. I basically agree with @gordonwoodhull that .on('event', ...) is more idiomatic, so would be open to discussion of changing to that approach if @tannerlinsley agrees.

**Comment by tannerlinsley** [comment]
.on() is my vote. Some If not all of these features exist in universe. If
we're talking about separation of concerns, then my first guess would be
that this would be a universe thing. Tracking, saving, and restoring filter
state Is what it does best. I would lean towards keeping crossfilter
stateless and using it as the engine for universe. The on Change event is
fantastic. It supplies all of the info you'd need to do this, which is
exactly what universe does for you. Check it out!

**Comment by esjewett** [comment]
I'd agree with using Universe, but since Crossfilter is already storing these filters, I think it would make sense to provide a pretty dumb mechanism to return them?

One additional note: If we ever support native multi-value filters (e.g. Filter to 'A' or 'B', for which we currently need to use a filterFunction), this is going to pose a bit of a problem.

**Comment by tannerlinsley** [comment]
Universe tracks multi values as is and applies them under the hood with
functions right now. Going native would be cool too. If crossfilter wanted
to track this info, that's fine. All the necessary logic is In universe
right now but is easily portable.

**Comment by MarkHerhold** [comment]
It seems to be that crossfilter is inherently stateful as-is. I am just looking to make it the source of truth rather than tacking extra hacks outside of crossfilter.

@esjewett I thought of the multi-filter issue as well. It would be trivial to make everything in the proposal say "filters" and return an array as I had before.

**Comment by MarkHerhold** [comment]
@tannerlinsley going forward, universe could ask crossfilter what filters were applied. I don't think anything else would have to change.

**Comment by tannerlinsley** [comment]
I need to update the docs for universe. But it supports tracking exact,
ranges, custom functions and inclusive arrays (multi value). At any given
time you can see what's applied, change it, etc. Another value is in the
toggling capability. You can replace or toggle single arrays of values.

Yeah I would be okay moving what makes sense to crossfilter.

**Comment by esjewett** [comment]
@MarkHerhold No, that's not really what I meant. It would still be a single filter, but a "special" kind. Not worth worrying about for this use case. It is correct to design this to return a single filter, not an array of them.

**Comment by MarkHerhold** [comment]
@esjewett ah, gotacha. My apologies.

I updated the proposal to say `.on()` instead of `.subscribe()`. I admit I am not sure if we should invest more in .onChange() and document it or if we should go ahead with `.on()`. Thoughts?

**Comment by esjewett** [comment]
Sounds like @tannerlinsley is OK with changing `onChange` over to the `on` syntax, with callback functions assigned to particular events instead of a callback function that receives an event string. Is that correct?

Either way, the `onChange`/`on` and the `dimension.currentFilter`\`crossfilter.currentFilters` changes are separate.

**Comment by tannerlinsley** [comment]
I'm alright with that. On(event name, cb) with any appropriate payload
being sent through the callback.

**Comment by MarkHerhold** [comment]
Just to be clear about the state of this, I do not have time to work on this enhancement for the foreseeable future.

**Comment by thomasgillet** [comment]
Hello all,

Using crossfilter for some months now, I sometimes run into minor caveats like this issue. I usually fix that in some kind of home-made wrapper, but maybe it's time to contribute rather than fiddling around by myself.

So, after more than 1 year of inactivity, is this issue still valid? If yes I would be happy to (try to) write a PR for it. The `currentFilter` feature seems simple enough for a first try.

**Comment by esjewett** [comment]
@thomasgillet Please do. I'm able to review and merge pull requests if/when you send them. My calendar is opening up a bit in 2018 but I am still underwater enough that I'm not going to try to start any changes myself.

**Comment by thomasgillet** [comment]
Ok, I'll give it a try.
Since I only plan to work on the filter getters, not the events, I created the more specific issue #92.

**Comment by dahlbyk** [comment]
Should this have been closed by #93?

**Comment by gordonwoodhull** [comment]
Replaced by #92 and implemented in #93.

[Documented here](https://github.com/crossfilter/crossfilter/wiki/API-Reference#dimension_currentFilter).
